### PR TITLE
Commentate+Tidy FuzzTests

### PR DESCRIPTION
**Pull request by StachuDotNet**
- [x] ~consider file naming pattern. X.FuzzTests.fs is pretty length.~ going to ignore this for now
- [x] ~mention properties, fuzzers, etc.~ nevermind, link to fscheck docs is fine
- [x] ~reread fscheck docs and reconsider if our usage of FsCheck is reasonable. It feels like we could have simpler tests if we 'inverted' the tests a bit, but hard to describe so far.~ can of worms for another day
- [x] ensure everything is well /// commentated (top of file, per test, etc.) (edit: I didn't do _everything_ but it's good enough for now)
- [x] improve readme text
- [x] include notes about dependencies in readme
- [x] tell how to run tests in readme. filters. link to docs
- [x] mention that other project(s) reference this one for some generators
- [x] mention that the fuzztests aren't run in CI

### CODE REVIEW

File: fsharp-backend/tests/FuzzTests/Json.FuzzTests.fs [hunk 1]
```diff
@@ -149,10 +151,11 @@ module LibJwtJson =
 
     actual .=. expected
 
-  let roundtripV1 (dv : RT.Dval) : bool =
+
+  let roundtripV0 (dv : RT.Dval) : bool =
```
Comment (StachuDotNet) [inline fsharp-backend/tests/FuzzTests/Json.FuzzTests.fs:155]: Here I simply moved the v0 fn above the v1 fn, to be consistent with the test list at the end of the module - looks like more churn than it is.

File: fsharp-backend/tests/FuzzTests/OCamlInterop.FuzzTests.fs [hunk 2]
```diff
@@ -60,27 +60,28 @@ let isInteroperable
...
 
 type Generator =
+  static member SafeString() : Arbitrary<string> = Arb.fromGen (G.string ())
+
   static member Expr() =
     Arb.Default.Derive()
     |> Arb.filter (function
-      // characters are not yet supported in OCaml
+      // characters are not supported in OCaml
+      // CLEANUP can be removed once OCaml gone
```
Comment (StachuDotNet) [inline fsharp-backend/tests/FuzzTests/OCamlInterop.FuzzTests.fs:73]: maybe this is redundant; the `FuzzTests` project probably needs a full once-over once OCaml is "gone"

File: fsharp-backend/tests/FuzzTests/Utils.FuzzTests.fs [hunk 3]
```diff
@@ -43,17 +45,17 @@ let testProperty (name : string) (x : 'a) : Test =
 let testPropertyWithGenerator (typ : System.Type) (name : string) (x : 'a) : Test =
   testPropertyWithConfig (baseConfigWithGenerator typ) name x
 
-
+/// Lower-level generators
 module Generators =
-  let nonNullString (s : string) : bool = s <> null
```
Comment (StachuDotNet) [inline fsharp-backend/tests/FuzzTests/Utils.FuzzTests.fs:48]: was unused

File: fsharp-backend/tests/FuzzTests/NodaTime.FuzzTests.fs [hunk 4]
```diff
@@ -1,4 +1,5 @@
-module FuzzTests.Date
+/// FuzzTests around NodaTime dependency
+module FuzzTests.NodaTime
```
Comment (StachuDotNet) [inline fsharp-backend/tests/FuzzTests/NodaTime.FuzzTests.fs:2]: The only test here was related to NodaTime. I felt the file name implied that we were testing _Dark_ Dates, so adjusted title to reflect reality. Having a separate module/file for Dark dates might make sense at some point.

**Comment by StachuDotNet** [comment]
@pbiggar ready for a review.
Note: I found `tpwg` repeatedly distracting, so removed the shortcut across a number of files.
It makes the code a bit less terse, but I think that's worth the distraction-removal.

**Comment by StachuDotNet** [comment]
Not sure why, but FYI GitHub seems to think I'm in the middle of reviewing the PR, and my intended one-off comments aren't respondable to me in the UI.

**Comment by pbiggar** [comment]
> @pbiggar I noticed [this](https://github.com/darklang/dark/blob/abbdec1357520e728eeabf097bc52ac5f1b7d320/fsharp-backend/tests/FuzzTests/ExecutePureFunctions.FuzzTests.fs#L29-L33) while tidying;
Should I steal some portion of this to place here instead? If not, is there another simple fix you think would be appropriate, or should I extract this into a separate issue?

I think I deliberately inlined these, to discover what the rules were for each scenario (eg printing to json isn't going to be the same as equality comparison, or something). No problem copying code but I'd avoid trying to DRY it.

File: fsharp-backend/tests/FuzzTests/DvalRepr.FuzzTests.fs [hunk 5]
```diff
@@ -1,3 +1,5 @@
+/// Generators and FuzzTests to ensure consistent behaviour
+/// of Runtime DVals across OCaml and F# backends
```
Comment (pbiggar) [inline fsharp-backend/tests/FuzzTests/DvalRepr.FuzzTests.fs:2]:
```suggestion
/// of Runtime Dvals across OCaml and F# backends
```

File: fsharp-backend/tests/FuzzTests/DvalRepr.FuzzTests.fs [hunk 6]
```diff
@@ -14,14 +16,14 @@ module OCamlInterop = LibBackend.OCamlInterop
 module DvalReprExternal = LibExecution.DvalReprExternal
 module DvalReprInternal = LibExecution.DvalReprInternal
 
-let tpwg = testPropertyWithGenerator
-
+/// Ensure text representation of DVals meant to be read by Dark users
```
Comment (pbiggar) [inline fsharp-backend/tests/FuzzTests/DvalRepr.FuzzTests.fs:19]:
```suggestion
/// Ensure text representation of Dvals meant to be read by Dark users
```

File: fsharp-backend/tests/FuzzTests/DvalRepr.FuzzTests.fs [hunk 7]
```diff
@@ -30,17 +32,19 @@ module DeveloperRepr =
       |> Arb.filter (function
         | RT.DFnVal _ -> false
         | RT.DFloat 0.0 -> false
-        | RT.DFloat infinity -> false
+        | RT.DFloat infinity -> false // TODO I don't think this is doing what we want it to!!
```
[the remark this answers is not in the document]
Comment (pbiggar) [reply fsharp-backend/tests/FuzzTests/DvalRepr.FuzzTests.fs:35]: I think I deliberately inlined these, to discover what the rules were for each scenario (eg printing to json isn't going to be the same as equality comparison, or something). No problem copying code but I'd avoid trying to DRY it.

File: fsharp-backend/tests/FuzzTests/HttpClient.FuzzTests.fs [hunk 8]
```diff
@@ -40,27 +41,33 @@ type QueryStringGenerator =
     |> Gen.map (String.concat "&")
     |> Arb.fromGen
 
-
+/// Checks that a DVal is consistently converted
```
Comment (pbiggar) [inline fsharp-backend/tests/FuzzTests/HttpClient.FuzzTests.fs:44]:
```suggestion
/// Checks that a Dval is consistently converted
```

File: fsharp-backend/tests/FuzzTests/HttpClient.FuzzTests.fs [hunk 9]
```diff
@@ -40,27 +41,33 @@ type QueryStringGenerator =
...
+/// to a URL-safe string across OCaml and F# backends
 let dvalToUrlStringExn (l : List<string * RT.Dval>) : bool =
   let dv = RT.DObj(Map l)
 
   DvalReprExternal.toUrlString dv .=. (OCamlInterop.toUrlString dv).Result
 
+/// Checks that a DVal is consistently converted
```
Comment (pbiggar) [inline fsharp-backend/tests/FuzzTests/HttpClient.FuzzTests.fs:51]:
```suggestion
/// Checks that a Dval is consistently converted
```

File: fsharp-backend/tests/FuzzTests/HttpClient.FuzzTests.fs [hunk 10]
```diff
@@ -40,27 +41,33 @@ type QueryStringGenerator =
...
+/// to a querystring-safe string across OCaml and F# backends
 let dvalToQuery (l : List<string * RT.Dval>) : bool =
   let dv = RT.DObj(Map l)
   DvalReprExternal.toQuery dv |> Result.unwrapUnsafe
   .=. (OCamlInterop.dvalToQuery dv).Result
 
+/// Checks that a DVal is consistently converted
```
Comment (pbiggar) [inline fsharp-backend/tests/FuzzTests/HttpClient.FuzzTests.fs:58]:
```suggestion
/// Checks that a Dval is consistently converted
```

File: fsharp-backend/tests/FuzzTests/Json.FuzzTests.fs [hunk 11]
```diff
@@ -20,9 +22,9 @@ module DvalReprExternal = LibExecution.DvalReprExternal
 module DvalReprInternal = LibExecution.DvalReprInternal
 module G = Generators
 
-let tpwg = testPropertyWithGenerator
-
-
+/// Generators and FuzzTests ensuring Runtime DVals are
```
Comment (pbiggar) [inline fsharp-backend/tests/FuzzTests/Json.FuzzTests.fs:25]:
```suggestion
/// Generators and FuzzTests ensuring Runtime Dvals are
```

File: fsharp-backend/tests/FuzzTests/README.md [hunk 12]
```diff
@@ -1,4 +1,28 @@
...
+
+## Writing Tests
+
+[FsCheck](https://fscheck.github.io/FsCheck/) is used to generate test cases and encode properties that we expect to be true.
+
+## Running Tests
+
+[Expecto](https://github.com/haf/expecto) is used to run tests.
+
+You can trigger test runs by running `./scripts/run-fsharp-tests` within the dev container
```
Comment (pbiggar) [inline fsharp-backend/tests/FuzzTests/README.md:22]:
```suggestion
You can trigger test runs by running `./scripts/run-fsharp-fuzzer` within the dev container
```

**Review by pbiggar — CHANGES_REQUESTED** [review]
This is great, thanks!

**Review by pbiggar — APPROVED** [review]

**Pull request merged**